pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -63,21 +96,52 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
-	// TODO: call pre-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
-	followup, err := next.Execute(msg)
+
+	// trigger the event if the message type is registered (if it's not outbound)
+	eventsTriggered := false
+	if !msg.Outbound {
+		eventsTriggered, err = s.sendEvent(msg, thid, next)
+		if err != nil {
+			return fmt.Errorf("send events failed: %w", err)
+		}
+	}
+
+	// if no events are trigger continue the execution
+	if !eventsTriggered {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// Register events on DID Exchange protocol. The events are triggered for incoming ConnectionRequest, ConnectionResponse
+// or ConnectionAck message types.
+func (s *Service) Register(ch chan<- event.DIDCommEvent) error {
```

Comment:
How does a listener unregister?

Reply:
I was thinking about adding this in second cut as this PR is already more than 600 lines. For now added an [empty function](https://github.com/hyperledger/aries-framework-go/compare/5839b0f509b0932a67abd562fb6a05dc181f2e08..5dfa12128f1ba2867b1a8c7c47464e6f9b8c5b7a#diff-f761c91e753eba1c42a68076df06ce12R134) with a TODO.